server/builtin/app.go

```diff
@@ -0,0 +1,178 @@
+package builtin
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"path"
+
+	"github.com/mattermost/mattermost-plugin-apps/apps"
+	"github.com/mattermost/mattermost-plugin-apps/server/config"
+	"github.com/mattermost/mattermost-plugin-apps/server/httpout"
+	"github.com/mattermost/mattermost-plugin-apps/server/proxy"
+	"github.com/mattermost/mattermost-plugin-apps/server/store"
+	"github.com/mattermost/mattermost-plugin-apps/upstream"
+	"github.com/mattermost/mattermost-plugin-apps/utils"
+)
+
+const (
+	AppID          = "apps"
+	AppDisplayName = "Mattermost Apps plugin"
+	AppDescription = "Install and manage Mattermost Apps"
+)
+
+const (
+	fURL            = "url"
+	fConsent        = "consent"
+	fSecret         = "secret"
+	fAppID          = "app"
+	fVersion        = "version"
+	fIncludePlugins = "include_plugins"
+	fUserID         = "user"
+)
+
+const (
+	pDebugBindings      = "/debug-bindings"
+	pDebugClean         = "/debug-clean"
+	pInfo               = "/info"
+	pList               = "/list"
+	pUninstall          = "/uninstall"
+	pEnable             = "/enable"
+	pDisable            = "/disable"
+	pInstallURL         = "/install-url"
+	pInstallS3          = "/install-s3"
+	pInstallMarketplace = "/install-marketplace"
+	pInstallConsent     = "/install-consent"
+)
+
+type builtinApp struct {
+	conf    config.Service
+	proxy   proxy.Service
+	store   *store.Service
+	httpOut httpout.Service
+
+	router map[string]func(apps.CallRequest) apps.CallResponse
+}
+
+var _ upstream.Upstream = (*builtinApp)(nil)
+
+func NewBuiltinApp(conf config.Service, proxy proxy.Service, store *store.Service, httpOut httpout.Service) *builtinApp {
```

Comment:
Instead of a full fledged app, wouldn't have been easier to just hijack the "getBindings" and the "call" endpoints? This seems in one hand clearer (it is clear that the app framework is an app itself) but at the same time, the logical process feels a bit convoluted (to get my bindings, I call an endpoint of myself to get the bindings; to make a call, I call an endpoint of myself to get the call...). 0/5

As an example, the builtin commands in the server code do not go through the usual command flow.

Reply:
@larkox I actually first started defining "Upstream" in a more semantic way, with `Call`, `Notify`, `GetBindings`, etc. Then almost immediately it was apparent that the interface should present the minimum (simplest) required functionality, thus `Roundtrip`.

Interfacing with builtin Apps at the `Upstream` layer rather than at the `Proxy.callApp` layer makes sense to me - more uniform. The (builtin) app still needs a big switch case in its `Call` handler, so it's not saving much hooking into `callApp` and separating `GetBindings` etc into separate functions.